### PR TITLE
NPC Interactable w/ Mission Management & Progression

### PR DESCRIPTION
**Pull request by chadleytan**
Main Changes:
* Created `NPCInteractable` class which is a subclass of `DialougeInteractable.cs`
  * Loads conversation that NPC would have based on the progression of a mission
  * Must make and drag before, during, after and default conversations + mission prefab (if NPC provides mission) 
  * Have event listeners for progression of mission (e.g. objective complete) and call functions in `ProgressManager.cs` that keeps track of mission status
* For `ProgressManager.cs` added functionality to add missions and update each status. This is used so for dialogue of the NPC to know which conversation to load based on the status of the mission
* Made NurseNPC prefab with interactbox and speech bubble as children
* Froze player when conversing and prevented certain functionality in `OnInteract` to occur when trying to advance in a conversation

### CODE REVIEW

File: Assets/Scripts/Interactables/Dialogue/DialogueInteractable.cs [hunk 1]
```diff
@@ -9,17 +9,31 @@ public class DialogueInteractable : Interactable
...
+    public override void OnInteract()  
     {
-        speakerUIBirdie = birdie.GetComponentInChildren<SpeakerUI>();
-        speakerUINPC = GetComponentInChildren<SpeakerUI>();
-
-        HideInteractUI();
-        AdvanceConversation();
-
-        base.OnInteract(birdie);
+        if(!isConversing)
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Dialogue/DialogueInteractable.cs]: space after `if`, so `if (!..blah.)`

File: Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    public List<Conversation> defaultConvos;
+
+    public List<MissionNPC> missionsOffered;
+    private MissionNPC currentMissionNPC;                            // Current Mission given by this NPC (should only be 1 per NPC)
+    private AMission startedMission;                                 // started mission needed to end mission
+
+
+    public override void OnInteract()
+    {
+        if(!isConversing)
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs]: space after `if`

File: Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            // Load the conversation of the NPC based on mission progress
+            if (missionsOffered.Count == 0)
+            {
+                // Return a random default convo (Temporary until better system is set for default convos)
+                conversation = defaultConvos[Random.Range(0, defaultConvos.Count - 1)];
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs]: I think should just be `defaultConvos.Count` without the `-1`? 
`Random.Range` is like `[start, finish)` ya know
id test it out to make sure though

File: Assets/Scripts/Interactables/Interactable.cs [hunk 4]
```diff
@@ -5,7 +5,8 @@
 
 public class Interactable : MonoBehaviour, IInteractable
 {
-    private GameObject player;
+    // TODO: player has already been assigned on trigger enter so remove passing it in for all functions
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Interactable.cs]: no more need for this TODO

File: Assets/Scripts/Interactables/Interactable.cs [hunk 5]
```diff
@@ -29,9 +30,9 @@ protected void Update()
             // User chooses to interact with the item
             if (Input.GetButtonDown(Constants.INPUT_INTERACTABLE_GETDOWN))
             {
-                FaceInteractable();
+                //FaceInteractable();
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Interactable.cs:31]: add a TODO comment to re-add this?

File: Assets/Scripts/Interactables/Interactable.cs [hunk 6]
```diff
@@ -93,12 +94,13 @@ public void FaceInteractable()
...
         // Possible issue to prevent another coroutine being called if player is already rotating
-        StartCoroutine(RotateAnimation(player, rotation, player.GetComponent<PlayerController>().turnSpeed));
+        player.GetComponent<PlayerController>().CanMove = false;
+        StartCoroutine(RotateAnimation(player, rotation, player.GetComponent<PlayerController>().turnSpeed, () => UnfreezePlayer()));
 
     }
 
     // Coroutine that animates the rotation of the given object to the desiredRotation at a set turn speed
-    private IEnumerator RotateAnimation(GameObject obj, Quaternion desiredRotation, float turnSpeed)
+    protected IEnumerator RotateAnimation(GameObject obj, Quaternion desiredRotation, float turnSpeed, Action onFinishCallback)
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Interactable.cs]: To make this function more general maybe we should see how to set a default null-ish value for `Action`, and just conditionally call it if its not null (so it's not necessary pass in a callback).

File: Assets/Scripts/ProgressManager.cs [hunk 7]
```diff
@@ -6,7 +6,8 @@ public class ProgressManager: Singleton<ProgressManager>
 {
     // Should communicate with Dialogue System on what dialogue should show when interacting with same interactable
     // Needed to communicate with the dialogue system
-    public List<InProgressMissionContainer> completedMissions;
+    //public List<IMission> Missions;
```
Comment (jurcaua) [inline Assets/Scripts/ProgressManager.cs]: can we delete this commented out thing now?

File: Assets/Scripts/ProgressManager.cs [hunk 8]
```diff
@@ -6,7 +6,8 @@ public class ProgressManager: Singleton<ProgressManager>
...
+    private Dictionary<AMission, MissionStatusCode> MissionsStatus;
```
Comment (jurcaua) [inline Assets/Scripts/ProgressManager.cs]: rename to `missionStatus` to keep lowercase first character convention

**Review by jurcaua — CHANGES_REQUESTED** [review]
we're so close
now u just have to pass me
<3

File: Assets/Scripts/Interactables/Interactable.cs (hunk 6 above)
Comment (chadleytan) [reply Assets/Scripts/Interactables/Interactable.cs]: Ok i'll make that a TODO for now if that's chill

Comment (jurcaua) [reply Assets/Scripts/Interactables/Interactable.cs]: check discord!!!

File: Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs (hunk 3 above)
Comment (chadleytan) [reply Assets/Scripts/Interactables/Dialogue/NPCDialogue/NPCInteractable.cs]: Ok yeah it min inclusive and max exclusive!

**Review by jurcaua — APPROVED** [review]
good to go!

**Pull request merged**